### PR TITLE
DeepMimic Gym enhancements

### PR DESCRIPTION
**Pull request by ManifoldFR**
This PR is a sequel to issue #2821 and PR #2822 where first fixes were made to the DeepMimic Gym environment (mostly the policy querying frequency, which should be 30Hz as in the paper). These fixes are finalized and further enhancements were made as said in issue #2868:
* the render camera tracks the character
* the actions and states are rescaled as Jason Peng's original code does (this can be disabled)
* I added a test mode flag that makes the `reset()` method reset the reference clip to the beginning instead of random, which is better for evaluation/visualization

I was able to train the backflip policy using my own PyTorch codebase (which does not rely on MPI): see [this video](https://youtu.be/K-fYcoDdP7M).
The sample efficiency was not as good as the initial TensorFlow code I think but it could just be me choosing bad hyperparameters.
As mentioned in issue #2871 the character's root position tends to drift from the reference motion.

Resolves issue #2868.

### CODE REVIEW

**Comment by erwincoumans** [comment]
Great, thanks for finishing the DeepMimic Gym environment, glad to see you can train it with another implementation.

**Comment by erwincoumans** [comment]
Do you have a link to your PyTorch training codebase?

Did you try training the env using stable_baselines?

Recently, Jason Peng implemented the motion_imitation env also trained using stable_baselines.
The curriculum improves sample efficiency, we don't enable this in our DeepMimic environment, I think:
https://github.com/google-research/motion_imitation/blob/246b2e786ccb2033cf0f4eff6fe6c62dbfb51eee/motion_imitation/envs/env_wrappers/imitation_wrapper_env.py#L26

**Comment by ManifoldFR** [comment]
My PyTorch codebase is currently private because it's part of an ongoing IL project with my institution, and I haven't tried training using another RL codebase yet.

I don't remember the concept of curriculum in the original animal imitation paper, though, but I'll take a look thanks !
Right now I'm training the cartwheel and crawling motions, so if I make further changes to the Gym env and add more options I will make another contribution.

**Comment by erwincoumans** [comment]
Thanks. Do you also make sure the original code still trains and 'enjoys' ? How does the training speed compare with your code?

python3 mpi_run.py --arg_file train_humanoid3d_walk_args.txt --num_workers 34
python3 -m pybullet_envs.deep_mimic.testrl --arg_file run_humanoid3d_walk_args.txt

**Comment by ManifoldFR** [comment]
I just checked and running the policies still works well. I didn't make any breaking changes to the sim environment file pybullet_deep_mimic_env.py and Jason Peng's RL code doesn't use the other file I changed.
I did not check if the training behaviour changed though.
If I remember another difference with the initial framework was the COM reward in `humanoid_stable_pd.HumanoidStablePD` had a weight of 0, right?

**Comment by tfederico** [comment]
I would also be interested in a PyTorch version. @ManifoldFR could you share an example of how to use it instead of MPI?

**Comment by ManifoldFR** [comment]
> I would also be interested in a PyTorch version. @ManifoldFR could you share an example of how to use it instead of MPI?

I think the closest public thing to what I did would be [stable-baselines3](/DLR-RM/stable-baselines3/). Since the env uses the Gym interface I can just instantiate a vectorized env that runs in subprocesses and I don't need MPI anymore. stable-baselines has its own implementation of vectorized envs, but the current version of OpenAI Gym also has these.

**Comment by tfederico** [comment]
thank you, I will take a look at the link you suggested!
